docs/infrastructure/logging.md

```diff
@@ -0,0 +1,243 @@
+---
+layout: developer-doc
+title: Logging Service
+category: infrastructure
+tags: [infrastructure, logging, debug]
+order: 6
+---
+
+# Logging
+
+A centralized service has been created to better manage logging from multiple
+components. This service can be started with one of the main components,
+allowing other components connect to it. All logs are gathered in one place for
+easier analysis of component's interaction.
+
+<!-- MarkdownTOC levels="2,3" autolink="true" -->
+
+- [Protocol](#protocol)
+  - [Types](#types)
+  - [Messages](#messages)
+  - [Examples](#examples)
+- [JVM Architecture](#jvm-architecture)
+  - [SLF4J Interface](#slf4j-interface)
+  - [Setting Up Logging](#setting-up-logging)
+
+<!-- /MarkdownTOC -->
+
+## Protocol
+
+The service is designed to be usable by a variety of components. It relies on a
+WebSocket connection to a specified endpoint that exchanges JSON-encoded text
+messages. The communication is uni-directional - the only messages are log
+messages that are sent from a connected client to the server that gathers all
+logs.
+
+### Types
+
+##### `LogLevel`
+
+The log level encoded as a number. Possible values are:
+
+- 0 - indicating `ERROR` level,
+- 1 - indicating `WARN` level,
+- 2 - indicating `INFO` level,
+- 3 - indicating `DEBUG` level,
+- 4 - indicating `TRACE` level.
+
+```typescript
+type LogLevel = number;
+```
+
+##### `UTCTime`
+
+Message timestamp encoded as milliseconds elapsed from the UNIX epoch, i.e.
+00:00:00 UTC, 1 January 1970.
```

Comment:
I thought that a UNIX timestamp will be easier to handle - it seems most portable as I think all tools can construct a datetime from such timestamp; and then the `Printer` can format it in any way it wants.
It will need to be parsed in the future when the reordering file printer is implemented and I though that it is simpler to parse a UNIX timestamp rather than a formatted datetime string as one has to keep what kind of formatting it uses etc.

But I guess changing it into a datetime is also a reasonable solution (we just have to choose a format and stick to it). So do you think I should keep the current solution or change it to a datetime?

Reply:
Using an actual datetime avoids any issues around epoch overflow and things like that. It's more robust in general.

Reply:
As discussed, we will leave the internal protocol at UNIX timestamps for simplicity.

The timestamp will by default be rendered in logging output as a UTC datetime with the zone marked.